astropy/io/misc/qdp.py

```diff
@@ -0,0 +1,451 @@
+import copy
+from itertools import groupby
+import numpy as np
+import warnings
+from astropy.table import Table
+from astropy.io import registry as io_registry
+
+
+def is_qdp(origin, filepath, fileobj, *args, **kwargs):
+    if filepath is not None:
+        return filepath.endswith(('.qdp'))
+    return False
+
+
+def line_type(line):
+    """Interpret a QDP file line
+
+    Parameters
+    ----------
+    line : str
+        a single line of the file
+
+    Returns
+    -------
+    type : str
+        Line type: "comment", "command", or "data"
+
+    Examples
+    --------
+    >>> line_type("READ SERR 3")
+    'command'
+    >>> line_type(" \\n    !some gibberish")
+    'comment'
+    >>> line_type("   ")
+    'comment'
+    >>> line_type(" 21345.45")
+    'data,1'
+    >>> line_type(" 21345.45 NO")
+    'data,2'
+    >>> line_type(" 21345.45 ! a comment to disturb")
+    'data,1'
+    >>> line_type("NO NO NO NO NO")
+    'new'
+    >>> line_type(" some non-comment gibberish")
+    Traceback (most recent call last):
+        ...
+    ValueError: Unrecognized QDP line...
+    """
+    line = line.strip()
+    if line == "":
+        return "comment"
+
+    # Look for in-line comments (In lines containing other data or commands,
+    # not starting with those comments)
+    comment_in_line = line.find("!")
+    if comment_in_line > 1:
+        line = line[:comment_in_line]
+
+    probe = line.split()[0]
+    n = len(line.split())
+
+    if probe.startswith("!"):
+        return "comment"
+
+    # Full line of NOs -> new table
+    if probe.strip("NO ") == "":
```

Comment:
I'm impressed by the `_line_type_re`, but if I may, I had a hard time following that up :). 
I thought about using re's but decided otherwise because I wanted something easy to read and fix. No strong opinions.

Just one comment about the `_decimal_re`: does it understand numbers in scientific format?

Reply:
However, I tried all doctests and this solution doesn't pass them. It fails at the very first:
```
103     Examples
104     --------
105     >>> line_type("READ SERR 3")
UNEXPECTED EXCEPTION: ValueError('Unrecognized QDP line: READ SERR 3')
```
As per the "N O N NOON OON O" line... I hope nobody is so evil to write something like that :D

Reply:
Ok, I expanded your regular expression and now it works in all cases, including scientific notation!